WooCommerce/Classes/Authentication/Store Creation/StoreCreationStoreSwitchScheduler.swift

```diff
@@ -0,0 +1,97 @@
+import Foundation
+import Yosemite
+import Combine
+
+protocol StoreCreationStoreSwitchScheduler {
+    func savePendingStoreSwitch(siteID: Int64, expectedStoreName: String)
+
+    func removePendingStoreSwitch()
+
+    var isPendingStoreSwitch: Bool { get }
+
+    func listenToPendingStoreAndReturnSiteIDOnceReady() async throws -> Int64?
+}
+
+/// Stores store creation info and informs once the store is ready
+///
+final class DefaultStoreCreationStoreSwitchScheduler: StoreCreationStoreSwitchScheduler {
+    private let stores: StoresManager
+    private let userDefaults: UserDefaults
+    private var storeStatusChecker: StoreCreationStatusChecker?
+    private var jetpackSiteSubscription: AnyCancellable?
+
+    private var siteIDPendingStoreSwitch: Int64? {
+        set {
+            userDefaults[.siteIDPendingStoreSwitch] = newValue
+        }
+        get {
+            userDefaults[.siteIDPendingStoreSwitch] as? Int64
+        }
+    }
+
+    private var expectedStoreNamePendingStoreSwitch: String? {
+        set {
+            userDefaults[.expectedStoreNamePendingStoreSwitch] = newValue
+        }
+        get {
+            userDefaults[.expectedStoreNamePendingStoreSwitch] as? String
+        }
+    }
+
+    init(stores: StoresManager = ServiceLocator.stores,
+         userDefaults: UserDefaults = .standard,
+         storeStatusChecker: StoreCreationStatusChecker? = nil) {
+        self.stores = stores
+        self.userDefaults = userDefaults
+        self.storeStatusChecker = storeStatusChecker
+    }
+
+    func savePendingStoreSwitch(siteID: Int64, expectedStoreName: String) {
+        siteIDPendingStoreSwitch = siteID
+        expectedStoreNamePendingStoreSwitch = expectedStoreName
+    }
+
+    func removePendingStoreSwitch() {
+        siteIDPendingStoreSwitch = nil
+        expectedStoreNamePendingStoreSwitch = nil
+    }
+
+    var isPendingStoreSwitch: Bool {
+        guard siteIDPendingStoreSwitch != stores.sessionManager.defaultStoreID else {
+            siteIDPendingStoreSwitch = nil
+            expectedStoreNamePendingStoreSwitch = nil
+            return false
+        }
+
+        return siteIDPendingStoreSwitch != nil && expectedStoreNamePendingStoreSwitch != nil
+    }
+
+    @MainActor
+    func listenToPendingStoreAndReturnSiteIDOnceReady() async throws -> Int64? {
+        guard let siteID = userDefaults[.siteIDPendingStoreSwitch] as? Int64,
+              let expectedStoreName = userDefaults[.expectedStoreNamePendingStoreSwitch] as? String else {
+            return nil
+        }
+
+        let statusChecker = storeStatusChecker ?? DefaultStoreCreationStatusChecker(isFreeTrialCreation: true, storeName: expectedStoreName, stores: stores)
```

Comment:
Is it possible that the existing `storeStatusChecker` has a different store name? Should we reassign the checker or compare the store name to be safe.

Reply:
Good point. I injected this to avoid a delay in unit tests. I solved this by injecting the retry time interval instead in this commit 4bc67b28381be8e7a4dc1335acc23c94ce99fed1